src/jit/optimizer.cpp

```diff
@@ -1416,410 +1418,1018 @@ void Compiler::optCheckPreds()
+            else
+            {
+                BlockSetOps::AddElemD(comp, oldBlocksInLoop, blockNum);
+            }
+        }
 
-    for (head = fgFirstBB; head->bbNext; head = head->bbNext)
+        bool testAndInsert(unsigned int blockNum)
+        {
+            if (blockNum > oldBlockMaxNum)
+            {
+                unsigned int shiftedNum = blockNum - oldBlockMaxNum;
+                if (!BlockSetOps::IsMember(comp, newBlocksInLoop, shiftedNum))
+                {
+                    BlockSetOps::AddElemD(comp, newBlocksInLoop, shiftedNum);
+                    return false;
+                }
+            }
+            else
+            {
+                if (!BlockSetOps::IsMember(comp, oldBlocksInLoop, blockNum))
+                {
+                    BlockSetOps::AddElemD(comp, oldBlocksInLoop, blockNum);
+                    return false;
+                }
+            }
+            return true;
+        }
+    };
+
+    LoopBlockSet loopBlocks; // Set of blocks identified as part of the loop
+    Compiler*    comp;
+
+    // See LoopSearch class comment header for a diagram relating these fields:
+    BasicBlock* head;   // Predecessor of unique entry edge
+    BasicBlock* first;  // Lexically first in-loop block
+    BasicBlock* top;    // Successor of back-edge from BOTTOM
+    BasicBlock* bottom; // Predecessor of back-edge to TOP, also lexically last in-loop block
+    BasicBlock* entry;  // Successor of unique entry edge
+
+    BasicBlock*   lastExit;       // Most recently discovered exit block
+    unsigned char exitCount;      // Number of discovered exit edges
+    unsigned int  oldBlockMaxNum; // Used to identify new blocks created during compaction
+    BlockSet      bottomBlocks;   // BOTTOM blocks of already-recorded loops
+#ifdef DEBUG
+    bool forgotExit = false; // Flags a rare case where lastExit gets nulled out, for assertions
+#endif
+
+public:
+    bool changedFlowGraph = false; // Signals to caller that fgUpdateChangedFlowGraph needs to be called
+
+    LoopSearch(Compiler* comp)
+        : loopBlocks(comp), comp(comp), oldBlockMaxNum(comp->fgBBNumMax), bottomBlocks(BlockSetOps::MakeEmpty(comp))
     {
-        top       = head->bbNext;
-        exit      = nullptr;
-        exitCount = 0;
+        // Make sure we've renumbered such that the bitsets can hold all the bits
+        assert(comp->fgBBNumMax <= comp->fgCurBBEpochSize);
+    }
+
+    //------------------------------------------------------------------------
+    // findLoop: Search for a loop with the given HEAD block.
+    //
+    // Arguments:
+    //    head - Block to be the HEAD of any loop identified
+    //
+    // Return Value:
+    //    true  - Found a valid loop.
+    //    false - Did not find a valid loop.
+    //
+    // Notes:
+    //    May modify flow graph go make loop compact before returning.
+    //    Will set instance fields to track loop's extent and exits.
+    //
+    bool findLoop(BasicBlock* head)
```

Comment:
Fixed in https://github.com/dotnet/coreclr/pull/13314/commits/71989abf833df501ccbad9fb5ff42b6693749581